api/python/t4/packages.py

```diff
@@ -527,6 +527,13 @@ def push(self, name, path):
             A new package that points to the copied objects
         """
         dest = fix_url(path)
+        if dest.startswith('file://'):
+            # local push
+            os.makedirs(path, exist_ok=True)
+            pkg = self._materialize(dest)
+            pkg.build(name)
```

Comment:
so, my 1 concern here is that `build` and `push` implement the same logic wrt writing to the `package` and `named_packages` dirs, so it would be nice if this was a single function incase this logic needs to be updated/maintained (which it probably will for supporting arbitrary registries/dedupe etc).

I think the only reason it can't be just drop in replaced is that one version uses a tempfile and the other just uses a file handle locally. I think its ok to use a tempfile in both cases, although not ideal.

Refactoring that means you don't need to special case file and s3 in this method at all.